### PR TITLE
Crier slack reporter: use pointers for merging configs

### PR DESCRIPTION
**Pull request by chaodaiG**
This makes slack configuration using the same defaulting logic as other prow configurations. Will allow negating global slack configuration in job config.

### CODE REVIEW

**Comment by chaodaiG** [comment]
/cc @cjwagner @alvaroaleman 
cc @howardjohn

File: prow/apis/prowjobs/v1/types.go [hunk 1]
```diff
@@ -286,10 +286,39 @@ type ReporterConfig struct {
...
-	Channel           string         `json:"channel,omitempty"`
-	JobStatesToReport []ProwJobState `json:"job_states_to_report,omitempty"`
-	ReportTemplate    string         `json:"report_template,omitempty"`
+	Host              *string         `json:"host,omitempty"`
+	Channel           *string         `json:"channel,omitempty"`
+	JobStatesToReport *[]ProwJobState `json:"job_states_to_report,omitempty"`
+	ReportTemplate    *string         `json:"report_template,omitempty"`
+}
+
+func (src *SlackReporterConfig) ApplyDefault(def *SlackReporterConfig) *SlackReporterConfig {
```
Comment (alvaroaleman) [inline prow/apis/prowjobs/v1/types.go:295]: Can we add a fuzztest that starts failing if someone adds a new field to `SlackReporterConfig` without adding merge logic, like we have for the decoration utils merging?

File: prow/config/config.go [hunk 2]
```diff
@@ -1042,11 +1042,8 @@ type ManagedWebhooks struct {
 // SlackReporter represents the config for the Slack reporter. The channel can be overridden
 // on the job via the .reporter_config.slack.channel property
 type SlackReporter struct {
-	JobTypesToReport  []prowapi.ProwJobType  `json:"job_types_to_report,omitempty"`
-	JobStatesToReport []prowapi.ProwJobState `json:"job_states_to_report,omitempty"`
-	Host              string                 `json:"host,omitempty"`
-	Channel           string                 `json:"channel"`
-	ReportTemplate    string                 `json:"report_template"`
+	JobTypesToReport *[]prowapi.ProwJobType `json:"job_types_to_report,omitempty"`
+	prowapi.SlackReporterConfig
```
Comment (alvaroaleman) [inline prow/config/config.go:1046]:
```suggestion
	prowapi.SlackReporterConfig `json:",inline"`
```

File: prow/apis/prowjobs/v1/types.go [hunk 3]
```diff
@@ -286,10 +286,39 @@ type ReporterConfig struct {
 }
 
 type SlackReporterConfig struct {
-	Host              string         `json:"host,omitempty"`
-	Channel           string         `json:"channel,omitempty"`
-	JobStatesToReport []ProwJobState `json:"job_states_to_report,omitempty"`
-	ReportTemplate    string         `json:"report_template,omitempty"`
+	Host              *string         `json:"host,omitempty"`
+	Channel           *string         `json:"channel,omitempty"`
+	JobStatesToReport *[]ProwJobState `json:"job_states_to_report,omitempty"`
```
Comment (alvaroaleman) [inline prow/apis/prowjobs/v1/types.go:291]: Slices are already pointer types, I don't think there is a need to use a double pointer

File: prow/apis/prowjobs/v1/types.go [hunk 4]
```diff
@@ -286,10 +286,39 @@ type ReporterConfig struct {
 }
 
 type SlackReporterConfig struct {
-	Host              string         `json:"host,omitempty"`
-	Channel           string         `json:"channel,omitempty"`
-	JobStatesToReport []ProwJobState `json:"job_states_to_report,omitempty"`
-	ReportTemplate    string         `json:"report_template,omitempty"`
+	Host              *string         `json:"host,omitempty"`
```
Comment (alvaroaleman) [inline prow/apis/prowjobs/v1/types.go:289]: Why do these need to be string pointers? Is the empty string a valid config for any of  them? If not, just merging when its empty would make things easier IMHO

File: prow/crier/reporters/slack/reporter_test.go [hunk 5]
```diff
@@ -82,10 +91,34 @@ func TestShouldReport(t *testing.T) {
 			expected: true,
 		},
 		{
-			name: "Successful Job should not report",
+			name: "nil job config setting negate global",
```
Comment (cjwagner) [inline prow/crier/reporters/slack/reporter_test.go:94]: nit: s/nil/empty/

**Review by cjwagner — COMMENTED** [review]
Thanks for fixing this up Chao, this seems much better than the previous config handling.

File: prow/apis/prowjobs/v1/types.go (hunk 1 above)
Comment (chaodaiG) [reply prow/apis/prowjobs/v1/types.go:295]: Added

File: prow/config/config.go [hunk 6]
```diff
@@ -1071,16 +1068,17 @@ func (cfg SlackReporterConfigs) GetSlackReporter(refs *prowapi.Refs) SlackReport
 
 func (cfg *SlackReporter) DefaultAndValidate() error {
```
[the remark this answers is not in the document]
Comment (chaodaiG) [reply prow/config/config.go]: Good point, will think about this later

**Review by chaodaiG — COMMENTED** [review]
Updated

File: prow/config/config.go (hunk 2 above)
Comment (chaodaiG) [reply prow/config/config.go]: done

File: prow/apis/prowjobs/v1/types.go (hunk 3 above)
Comment (chaodaiG) [reply prow/apis/prowjobs/v1/types.go]: good point, fixed

File: prow/apis/prowjobs/v1/types.go (hunk 4 above)
Comment (chaodaiG) [reply prow/apis/prowjobs/v1/types.go]: changed them back to string

File: prow/apis/prowjobs/v1/types.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (chaodaiG) [reply prow/apis/prowjobs/v1/types.go]: fixed

File: prow/crier/reporters/slack/reporter.go [hunk 7]
```diff
@@ -121,36 +117,38 @@ func (sr *slackReporter) GetName() string {
...
 	// Prow config.
-	jobStatesToReport := prowCfg.JobStatesToReport
-	if jobCfg != nil && len(jobCfg.JobStatesToReport) != 0 {
-		jobStatesToReport = jobCfg.JobStatesToReport
+	var allowedJobStates []v1.ProwJobState
+	if globalSlackConfig != nil && globalSlackConfig.JobStatesToReport != nil {
+		allowedJobStates = *globalSlackConfig.JobStatesToReport
+	}
+	if jobSlackConfig != nil && jobSlackConfig.JobStatesToReport != nil {
+		allowedJobStates = *jobSlackConfig.JobStatesToReport
```
[the remark this answers is not in the document]
Comment (chaodaiG) [reply prow/crier/reporters/slack/reporter.go]: done

File: prow/crier/reporters/slack/reporter_test.go (hunk 5 above)
Comment (chaodaiG) [reply prow/crier/reporters/slack/reporter_test.go]: done

**Review by cjwagner — APPROVED** [review]

**Comment by chaodaiG** [comment]
I don't think I have touched this test
/retest

**Pull request merged**